Fixes for building on ARM Macs

cfstras opened the pull request:
## Summary
I've been working on getting Mumble to build natively on ARM Macs (testing on an M1 MacBook Air.)

I've gotten Mumble and murmur to build, and everything except echo cancellation and shortcuts seems to work. (see caveats)

## Commit Explanations

Detailed explanations are also in the commit messages.

- Some fixes to make `-Doverlay=OFF` compile cleanly on macOS:
  - FIX(overlay): add missing ifdefs for USE_OVERLAY  …
  - FIX(overlay, mac): handle hideEvent without USE_OVERLAY  …
- Fix missing permissions
  - FEAT(mumble.plist, mac): add NSMicrophoneUsageDescription  …
- Replace usage of deprecated (and now disabled) APIs
  -  FEAT(audio, mac): migrate to Audio Component API  …
- FIX(build): disable overlay on ARM macs  …

## Caveats

- CMake only seems to search for libraries in `/usr/local/opt`.  
   Homebrew installs everything into `/opt/homebrew/opt` on aarch64,
   my current solution is to **completely remove** any x86_64 homebrew from /usr/local, and symlink `/usr/local/opt` to `/opt/homebrew/opt`. 
- I've added a CMake check that disables `overlay` on ARM macs, since https://github.com/mumble-voip/mach_override doesn't support ARM at all. 
- I have to build with `-Dice=off` right now, which just seems to be a problem of CMake not finding the library without additional trickery.
- Shortcuts do not work yet.
  - Haven't figured it out, it breaks with `GlobalShortcutMac: Unable to create EventTap. Global Shortcuts will not be available.`
- When I try to enable the echo cancellation introduced in #4694, I get a weird loop opening 
- I have to build with Xcode instead of Make, otherwise I get a crash on start (something about UCKeyTranslate getting an invalid pointer)

I used this script to make a build: https://gist.github.com/cfstras/f741087c77a899d5f77eb4664c5b4a70
Warning: this script tries to make the above mentioned link. This will probably cause problems if you already have an x86_64 homebrew installation. I recommend to nuke that first.

## Review Notes

One thing about c4a2930: in the `g.ocIntercept == null` case, it looks like `[event release]` is not called. is this a memory leak?
Or are we certain that the method is not even called in that case? If yes, then why is that guard even there?
Should there be an `[event release]` here?
https://github.com/mumble-voip/mumble/blob/master/src/mumble/GlobalShortcut_macx.mm#L230-L235

## Checks

- [x] My commits follow the [commit guidelines](https://github.com/mumble-voip/mumble/blob/master/COMMIT_GUIDELINES.md)

davidebeatrici commented:
Awesome, thank you very much for your contribution!

Could you fix the whitespace in ecc6ac67f1b0fece8a764f255f0e02054ccf075e, please?

Aside from that, everything looks good to me.

Krzmbrzl reviewed (commented):
> Should there be an [event release] here?
https://github.com/mumble-voip/mumble/blob/master/src/mumble/GlobalShortcut_macx.mm#L230-L235

I don't know the Mac internals but given that we use a release at the end of this function, I think we should have one here as well, yes.

Other than the whitespace issues that Davide already meantioned this LGTM.

Krzmbrzl commented:
@TerryGeng could you also have a look at these changes, please? You have much more experience with the macOS related code than we do.

TerryGeng commented:
@Krzmbrzl ARM Mac was just released months ago and I don't have one yet, so I can't really test it on my own. Sorry about that. But I did check the audio code and it looked good to me.

@cfstras Regarding the strange loop when enabling echo cancellation, can you provide more details?

Krzmbrzl commented:
> ARM Mac was just released months ago and I don't have one yet, so I can't really test it on my own. Sorry about that. But I did check the audio code and it looked good to me.

I only intended for you to have a look at the code as you might spot issues that we miss due to your experience with macOS. But given that you did not find any that's probably a good sign.
Thanks!

cfstras commented on CMakeLists.txt:
```diff
@@ -111,6 +111,12 @@ if(g15 AND WIN32)
+if(APPLE AND CMAKE_SYSTEM_PROCESSOR MATCHES "arm64")
+	# mach_override doesn't support ARM
+	# https://github.com/rentzsch/mach_override/issues/6
+	set(overlay OFF)
```
added and rebased.

cfstras commented:
Added the suggestions, fixed the whitespaces, and rebased.

@TerryGeng I'll test around some more with the loop problem and create a separate issue when I have details.

cfstras commented:
> Is this the `[event release]` that you are concerned with in terms of a possible memory leak?

Yes. also added that.

Krzmbrzl commented:
Thank you very much for your contribution! :+1:

josefleventon commented:
How do you go about building with Xcode @cfstras?

cfstras commented:
just add „-G Xcode“ as shown in
https://gist.github.com/cfstras/f741087c77a899d5f77eb4664c5b4a70 ; open
„build“ with xcode, and hit build. Edit the run config to start mumble.
That’s it.

Josef Leventon ***@***.***> schrieb am Fr. 19. März 2021 um
00:24:

> How do you go about building with Xcode @cfstras
> <https://github.com/cfstras>?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mumble-voip/mumble/pull/4765#issuecomment-802380903>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI2OAESF3LAFEFOOFYQT4TTEKDRPANCNFSM4XVKES4Q>
> .
>
-- 
—
Claus F. Strasburger
📱 +49 1520 1782213
✉️ ***@***.***

josefleventon commented:
I don't understand what you mean by "run config..."

cfstras commented:
Click on the ALL_BUILD, then select "Edit Scheme", and select "Mumble.app" in "Executable".
![image](https://user-images.githubusercontent.com/1156864/111819752-03573c80-88e1-11eb-9fbc-3df06e55c9e1.png)
![image](https://user-images.githubusercontent.com/1156864/111819775-09e5b400-88e1-11eb-824a-cc4d856472c8.png)
![image](https://user-images.githubusercontent.com/1156864/111819820-1538df80-88e1-11eb-8ed7-0d85ca56f028.png)
